### PR TITLE
 bandage box mechanic

### PR DESCRIPTION
**Pull request by bioastroiner**
adding features to bandage box
the box consists of 6 inventory slots for storing various bandages

- [x] ContainerClass
- [x] TileEntityClass
- [x] OpenGUIlogic
- [x] ScreenClass
- [x] InventoryLogic(only allows certain items to be placed in)

### CODE REVIEW

**Comment by bioastroiner** [comment]
should i make this PR a draft? cause i am not done yet.

**Comment by xf8b** [comment]
uhh your pr is just changing the imports. is this intended or did you mess up while committing?

**Comment by bioastroiner** [comment]
> uhh your pr is just changing the imports. is this intended or did you mess up while committing?

in the first commit 731c9ac there was still blueminecraftmods instead of teambluemods in the package name in all files, so i changed it and it affected some imports.

**Comment by bioastroiner** [comment]
> 
> 
> uhh your pr is just changing the imports. is this intended or did you mess up while committing?

ooooooooh holy shit
broh
i have no idea why it just commited just the import

**Comment by bioastroiner** [comment]
i think i actulley reverted some of my commit
broh some code is gone

**Review by zAG6z — DISMISSED** [review]
seems alright to me i think

**Comment by zAG6z** [comment]
oh shit i pressed somethin

**Comment by bioastroiner** [comment]
okay checklists tiked
only Binding Screen to Container Bruhcode witch i donno its kotling, maybe few twaeks to screen class

File: src/main/kotlin/io/github/teambluemods/healthmod/client/HealthModClient.kt [hunk 1]
```diff
@@ -38,5 +38,7 @@ object HealthModClient {
             HealthMod.rl("blood"),
             SyringeItem.BLOOD_PROPERTY
         )
+        //FIXME i have no idea how to BandageBoxScreen::new in freaking kotling, this lang is not like java
```
Comment (TheOnlyTails) [inline src/main/kotlin/io/github/teambluemods/healthmod/client/HealthModClient.kt:41]: `{ BandageBoxScreen() }`

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 2]
```diff
@@ -17,18 +17,71 @@
...
+        playerEntity: PlayerEntity,
+        handIn: Hand,
+        hit: BlockRayTraceResult
+    ): ActionResultType {
+        if (!worldIn.isRemote) {
+            val tile = worldIn.getTileEntity(pos)
+            if (tile is BandageBoxTileEntity) {
+                NetworkHooks.openGui(
+                    player as ServerPlayerEntity?,
+                    tile as BandageBoxTileEntity?,
```
Comment (TheOnlyTails) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:60]: You don't need to cast here. smart casting is a thing in kotlin.

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 3]
```diff
@@ -17,18 +17,71 @@
...
+        newState: BlockState,
+        isMoving: Boolean
+    ) {
+        if (state.block !== newState.block) {
+            val te = worldIn.getTileEntity(pos)
+            if (te is BandageBoxTileEntity) {
+                InventoryHelper.dropItems(
+                    worldIn,
+                    pos,
+                    (te as BandageBoxTileEntity?)?.items
```
Comment (TheOnlyTails) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:82]: See previous comment.

**Review by TheOnlyTails — CHANGES_REQUESTED** [review]
Some changes are still needed. Also, figure out why the checks are failing.

File: src/main/kotlin/io/github/teambluemods/healthmod/client/HealthModClient.kt (hunk 1 above)
Comment (xf8b) [reply src/main/kotlin/io/github/teambluemods/healthmod/client/HealthModClient.kt:41]: ::BandageBoxScreen

**Comment by xf8b** [comment]
wtf why are you requesting everyone, you only need 2 people to review

File: src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt [hunk 4]
```diff
@@ -17,10 +17,72 @@
...
 // TODO
-class BandageBoxScreen(titleIn: ITextComponent) : Screen(titleIn)
\ No newline at end of file
+@OnlyIn(Dist.CLIENT)
+class BandageBoxScreen(
+    screenContainer: BandageBoxContainer,
+    inventory: PlayerInventory,
+    titleIn: ITextComponent
+) :
+    ContainerScreen<BandageBoxContainer>(screenContainer, inventory, titleIn) {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt:41]: this is not how you do constructors

File: src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt [hunk 5]
```diff
@@ -17,10 +17,72 @@
...
+
+    init {
+        this.guiLeft = 0
+        this.guiTop = 0
+        this.xSize = 175
+        this.ySize = 183
+    }
+
+//    constructor() : this(ContainerTypeRegistries.BANDAGE_BOX.create(10,inventory),inventory = ) {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt:50]:
```kt
constructor(inventory: PlayerInventory, title: ITextComponent) : this(
    ContainerTypeRegistries.BANDAGE_BOX.create(10, inventory), 
    inventory, 
    title
)
```

File: src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt [hunk 6]
```diff
@@ -17,10 +17,72 @@
...
+            8.0f,
+            90.0f,
+            500 // don't know what is number is
+        )
+    }
+
+    override fun drawGuiContainerBackgroundLayer(matrixStack: MatrixStack, partialTicks: Float, x: Int, y: Int) {
+        RenderSystem.color4f(1.0f, 1.0f, 1.0f, 1.0f)
+        minecraft!!.getTextureManager().bindTexture(TEXTURE)
+        val x: Int = (width - this.xSize) / 2
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt:84]: don't need to specify type. see style guide

File: src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt [hunk 7]
```diff
@@ -17,10 +17,72 @@
...
+        val y: Int = (height - this.ySize) / 2
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/client/screen/BandageBoxScreen.kt:85]: *See previous comment.*

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 8]
```diff
@@ -17,18 +17,71 @@
...
+        state: BlockState,
+        worldIn: World,
+        pos: BlockPos,
+        playerEntity: PlayerEntity,
+        handIn: Hand,
+        hit: BlockRayTraceResult
+    ): ActionResultType {
+        if (!worldIn.isRemote) {
+            val tile = worldIn.getTileEntity(pos)
+            if (tile is BandageBoxTileEntity) {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:57]: new line before if. see code guidelines

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt (hunk 2 above)
Comment (xf8b) [reply src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:60]: also no need to spread the args over multiple lines

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 9]
```diff
@@ -17,18 +17,71 @@
...
+                NetworkHooks.openGui(
+                    player as ServerPlayerEntity?,
+                    tile as BandageBoxTileEntity?,
+                    pos
+                )
+            }
+            return ActionResultType.SUCCESS
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:64]: new line before return. see code guidelines.

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 10]
```diff
@@ -17,18 +17,71 @@
...
+        }
+        return ActionResultType.FAIL
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:66]: *See previous comment.*

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 11]
```diff
@@ -17,18 +17,71 @@
...
+    }
+
+    override fun onReplaced(
+        state: BlockState,
+        worldIn: World,
+        pos: BlockPos?,
+        newState: BlockState,
+        isMoving: Boolean
+    ) {
+        if (state.block !== newState.block) {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:76]: this is == not .equals

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 12]
```diff
@@ -17,18 +17,71 @@
...
+            val te = worldIn.getTileEntity(pos)
+            if (te is BandageBoxTileEntity) {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:78]: new line before if. see code guidelines.

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt [hunk 13]
```diff
@@ -17,18 +17,71 @@
...
+
+    override fun onReplaced(
+        state: BlockState,
+        worldIn: World,
+        pos: BlockPos?,
+        newState: BlockState,
+        isMoving: Boolean
+    ) {
+        if (state.block !== newState.block) {
+            val te = worldIn.getTileEntity(pos)
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:77]: descriptive names. see code guidelines

File: src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt [hunk 14]
```diff
@@ -17,34 +17,133 @@
...
+        addSlotBox(playerInventory, 9, leftCol, topRow, 9, 18, 3, 18)
+
+        // Hotbar
+        topRow += 0 //TODO player Hotbar topRow coord
+        addSlotRange(playerInventory, 0, leftCol, topRow, 9, 18)
+    }
+
+    /**
+     * unlike [addSlotBox] this method adds only a single row of boxes
+     * @param dx is gap size
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt:80]: descriptive names. see code guidelines.

File: src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt [hunk 15]
```diff
@@ -17,34 +17,133 @@
...
+                18 //TODO texture gap
+            )
+        }
+        layoutPlayerInventorySlots(0, 0) //TODO enter values
+    }
+
+    private fun layoutPlayerInventorySlots(leftCol: Int, topRow: Int) {
+        // Player inventory
+        var topRow = topRow //TODO
+        addSlotBox(playerInventory, 9, leftCol, topRow, 9, 18, 3, 18)
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt:71]: leftColumn. see code guidelines

File: src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt [hunk 16]
```diff
@@ -17,34 +17,133 @@
...
+            index++
+        }
+        return index
+    }
+
+    /**
+     * this method makes a huge Box slot with different rows
+     * and collumns, like an Inventory Slots
+     * so it is infact consists of smaller [addSlotRange]s
+     * @param dx x gap
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt:99]: descriptive names. see code guidelines.

File: src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt [hunk 17]
```diff
@@ -17,34 +17,133 @@
...
+     * @param dy y gap
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt:100]: *See previous comment.*

File: src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt [hunk 18]
```diff
@@ -17,34 +17,133 @@
...
 
     override fun canInteractWith(playerIn: PlayerEntity) =
         isWithinUsableDistance(callable, playerIn, BlockRegistries.BANDAGE_BOX)
 
+    override fun transferStackInSlot(playerIn: PlayerEntity?, index: Int): ItemStack? {
+        var itemstack = ItemStack.EMPTY
+        val slot: Slot? = inventorySlots[index]
+        if (slot != null && slot.hasStack) {
+            val itemstack1: ItemStack = slot.stack
+            itemstack = itemstack1.copy()
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/container/BandageBoxContainer.kt:130]: camel case. see code guidelines

File: src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt [hunk 19]
```diff
@@ -38,14 +39,18 @@ import net.minecraft.world.IBlockReader
...
+import javax.annotation.Nonnull
+
 
 class BandageBoxTileEntity(type: TileEntityType<*> = TileEntityTypeRegistries.BANDAGE_BOX) :
     LockableLootTileEntity(type) {
     private val items = InvWrapper(this)
     private var numPlayersUsing = 0
     private var contents = NonNullList.withSize(6, ItemStack.EMPTY)
-    private var itemHandler = LazyOptional.of { items }
+    private val handler: ItemStackHandler = createHandler()
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt:52]: no need to specify type. see code guidelines

File: src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt [hunk 20]
```diff
@@ -124,6 +129,27 @@ class BandageBoxTileEntity(type: TileEntityType<*> = TileEntityTypeRegistries.BA
...
 
+    private fun createHandler(): ItemStackHandler {
+        return object : ItemStackHandler(1) {
+            override fun onContentsChanged(slot: Int) {
+                // To make sure the TE persists when the chunk is saved later we need to
+                // mark it dirty every time the item handler changes
+                markDirty()
+            }
+
+            override fun isItemValid(slot: Int, @Nonnull stack: ItemStack): Boolean {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt:140]: no need for nonnull. this is already not null

File: src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt [hunk 21]
```diff
@@ -124,6 +129,27 @@ class BandageBoxTileEntity(type: TileEntityType<*> = TileEntityTypeRegistries.BA
...
+                return stack.item === ItemRegistries.BANDAGE
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt:141]: use expression. see guide guidelines.

File: src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt [hunk 22]
```diff
@@ -124,6 +129,27 @@ class BandageBoxTileEntity(type: TileEntityType<*> = TileEntityTypeRegistries.BA
...
+            }
+
+            @Nonnull
+            override fun insertItem(slot: Int, @Nonnull stack: ItemStack, simulate: Boolean): ItemStack {
+                return if (stack.item !== ItemRegistries.BANDAGE) {
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/common/tileentities/BandageBoxTileEntity.kt:146]: use expression, remove {}. see code guidelines.

**Review by xf8b — CHANGES_REQUESTED** [review]
multiple code guideline violations. otherwise looks ok

File: src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt (hunk 11 above)
Comment (xf8b) [reply src/main/kotlin/io/github/teambluemods/healthmod/common/blocks/BandageBoxBlock.kt:76]: oh wait nvm

File: src/main/kotlin/io/github/teambluemods/healthmod/client/HealthModClient.kt [hunk 23]
```diff
@@ -38,5 +38,7 @@ object HealthModClient {
...
+        //ScreenManager.registerFactory(ContainerTypeRegistries.BANDAGE_BOX,BandageBoxScreen)
```
Comment (xf8b) [inline src/main/kotlin/io/github/teambluemods/healthmod/client/HealthModClient.kt:42]:
```suggestion
        ScreenManager.registerFactory(ContainerTypeRegistries.BANDAGE_BOX, ::BandageBoxScreen)
```

**Review by xf8b — DISMISSED** [review]
looks good i think

**Review by xf8b — APPROVED** [review]

**Review by saksham4106 — APPROVED** [review]
I feel that it is good

**Comment by KingRealzYT** [comment]
why did i get a email saying i have to review this? if i do
yea this seems good 
If i dont
Goodbye

**Comment by xf8b** [comment]
you did before but idk why bioast requested you

**Review by TheOnlyTails — APPROVED** [review]
Lgtm

**Comment by zAG6z** [comment]
ok

**Review by zAG6z — APPROVED** [review]

**Comment by KingRealzYT** [comment]
Guess I'm just **built different**

**Comment by bioastroiner** [comment]
> 
> 
> Guess I'm just **built different**

yea didn't you were play tester ooops

**Pull request merged**